### PR TITLE
CAGRA pad dataset for 128bit vectorized load

### LINKED ISSUE #1458: CAGRA support arbitrary `dim` (number of features)

### PR DESCRIPTION
**Pull request by tfeher**
This PR adds padding to the dataset (if necessary) to make reading any of its rows compatible with 128bit vectorized loads. This change also enables handling arbitrary number of input features (before this PR each row had to be at least 64bit aligned, which constrained the acceptable number of input features).

Fixes #1458.

With this change, it is sufficient to keep a single "load type" specialization for the search kernels, which shall cut the binary size by half (#1459).

### CODE REVIEW

**Comment by tfeher** [comment]
Marked as breaking change because it removes the `load_bit_length` search parameter.

**Comment by tfeher** [comment]
There is still a bug with the new tests, otherwise it is ready for review.

**Comment by enp1s0** [comment]
The solution of the compilation error is:

https://github.com/tfeher/raft/blob/fb175783a6cb6b9d0e56f0c18bd84c269fce9bb1/cpp/include/raft/neighbors/detail/cagra/search_multi_cta.cuh#L231
```diff
231:      dataset_size,
+         dataset_ld,
232:      result_buffer_size,
```

**Comment by enp1s0** [comment]
It would be better to support a padded dataset in the argument of `cagra::sort_knn_graph`. What do you think?

**Review by enp1s0 — COMMENTED** [review]
The changes look good to me.

One comment:
It would be better to support a padded dataset in the argument of `cagra::sort_knn_graph`. What do you think?

**Comment by cjnolet** [comment]
@enp1s0 please fix the conflicts and we can merge this in for release 23.06.

**Comment by tfeher** [comment]
I have fixed a bug in the serialization routine, and added more tests. @enp1s0 could you have a quick look again at the changes?

> It would be better to support a padded dataset in the argument of cagra::sort_knn_graph. What do you think?

I think it is a good suggestion, and it might be also useful to allow the constructor of index to accept padded dataset (i.e. strided mdspan). I would prefer to make these changes in a follow up PR.

**Comment by enp1s0** [comment]
Thank you, @tfeher, for fixing it and adding tests. The code looks good to me.

**Review by enp1s0 — APPROVED** [review]
Thank you, @tfeher, for fixing it and adding tests. The code looks good to me.

File: cpp/include/raft/neighbors/detail/cagra/cagra_serialize.cuh [hunk 1]
```diff
@@ -59,7 +60,19 @@ void serialize(raft::resources const& res, std::ostream& os, const index<T, IdxT
...
+  RAFT_CUDA_TRY(cudaMemcpy2DAsync(host_dataset.data_handle(),
+                                  sizeof(T) * host_dataset.extent(1),
+                                  dataset.data_handle(),
+                                  sizeof(T) * dataset.stride(0),
+                                  sizeof(T) * host_dataset.extent(1),
+                                  dataset.extent(0),
+                                  cudaMemcpyDefault,
+                                  resource::get_cuda_stream(res)));
+  resource::sync_stream(res);
+  serialize_mdspan(res, os, host_dataset.view());
```
Comment (cjnolet) [inline cpp/include/raft/neighbors/detail/cagra/cagra_serialize.cuh:75]: Mostly a side question, but are we still planning to remove the dataset from the serialization in a future change?

File: cpp/include/raft/neighbors/cagra_types.hpp [hunk 2]
```diff
@@ -179,15 +183,36 @@ struct index : ann::index {
...
-               dataset.size(),
-               resource::get_cuda_stream(res));
+    if (dataset_.extent(1) == dataset.extent(1)) {
+      raft::copy(dataset_.data_handle(),
+                 dataset.data_handle(),
+                 dataset.size(),
+                 resource::get_cuda_stream(res));
+    } else {
+      // copy with padding
+      RAFT_CUDA_TRY(cudaMemsetAsync(
```
Comment (cjnolet) [inline cpp/include/raft/neighbors/cagra_types.hpp:199]: If this is going to be a more common practice, I wonder if we should consider centralizing this somewhere eventually. Probably doesn't need to be done yet, or even in this PR, though.

File: cpp/include/raft/neighbors/detail/cagra/search_multi_cta.cuh [hunk 3]
```diff
@@ -569,7 +546,7 @@ struct search : public search_plan_impl<DATA_T, INDEX_T, DISTANCE_T> {
   ~search() {}
 
   void operator()(raft::resources const& res,
-                  raft::device_matrix_view<const DATA_T, INDEX_T, row_major> dataset,
+                  raft::device_matrix_view<const DATA_T, INDEX_T, layout_stride> dataset,
```
Comment (cjnolet) [inline cpp/include/raft/neighbors/detail/cagra/search_multi_cta.cuh:549]: Would there be any benefit to using a  [padded layout](https://github.com/rapidsai/raft/blob/567dfd74a5c092cec61de6b52ea239ea91b2ed46/cpp/include/raft/core/mdspan.hpp#L58) here or having an overload for it in the public API just to simplify the conversion?

**Review by cjnolet — APPROVED** [review]
LGTM. Just a couple questions and a suggestion for a future improvement overall. Since CAGRA is still experimental, we have some leeway for API changes.

**Pull request merged**